#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
df = pd.read_csv(r"D:\Навчання\Data Analytics\КУРС DATA ANALIST\python\DATA\supermarket_sales.csv")

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

In [14]:
df.groupby(by = ['Branch'])['Total'].sum()

Branch
A    106200.3705
B    106197.6720
C    110568.7065
Name: Total, dtype: float64

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [26]:
df.groupby(by = ['Branch', 'City'])['Total'].sum().sort_values(ascending=False).head(1)

Branch  City     
C       Naypyitaw    110568.7065
Name: Total, dtype: float64

Хоч філія C має найвищий прибуток (110,568.71), але різниця з філіями A та В не суттєва. Відкривати ще один магазин на основі цього результату не доцільно

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [24]:
invoices_by_category = pd.pivot_table(
    df, index = 'Branch', 
    columns = 'Product line', 
    values = 'Invoice ID', 
    aggfunc = 'count'
)
display(invoices_by_category)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [25]:
invoices_by_category['Electronic accessories'].sort_values(ascending=False).head(1)

Branch
A    60
Name: Electronic accessories, dtype: int64

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


### Гіпотези:
1. Яка форма оплати (Cash, Credit card, Ewallet) приносить найбільше доходу?
2. Чи залежить середній чек від типу покупця (Customer type) і статі (Gender)?
3. У який день тижня відбувається найбільше покупок?

In [29]:
# 1. Яка форма оплати (Cash, Credit card, Ewallet) приносить найбільше доходу?
df.groupby('Payment')['Total'].sum().sort_values(ascending=False)

Payment
Cash           112206.570
Ewallet        109993.107
Credit card    100767.072
Name: Total, dtype: float64

Бачимо, що готівкові розрахунки в торгівельній мережі переважають.  

In [30]:
# 2. Чи залежить середній чек від типу покупця (Customer type) і статі (Gender)?
df.pivot_table(values='Total', index='Customer type', columns='Gender', aggfunc='mean')

Gender             Female        Male
Customer type                        
Member         337.727753  316.985419
Normal         332.233256  305.047581

З розрахунків бачимо, що найбільше купують жінки, а також, що середній чек більше у Members як серед чоловіків, так і серед жінок

In [36]:
# 3. У який день тижня відбувається найбільше покупок? 
df['Day'] = pd.to_datetime(df['Date']).dt.day_name()
df['Day'].value_counts()

Day
Saturday     164
Tuesday      158
Wednesday    143
Friday       139
Thursday     138
Sunday       133
Monday       125
Name: count, dtype: int64

Найбільше продажів приприпадає на суботу, найменше - на понеділок.  
Хоч ми цей функціонал не вивчали, мені здалось доцільним дослідити саме це питання, це було б досить логічно для реального проекту

### Використання для прийняття бізнес рішень
1. Просувати найвигідніший спосіб оплати (наприклад, додатковим кешбеком), або ж навпаки - впровадити заходи для збільшення оплат іншим способом. Як приклад, збільшення безготівкових розрахунків, щоб оптимізувати роботу касирів, впровадити каси самообслуговування тощо.
2. Створювати персоналізовані пропозиції за статтю покупця, спрямовувати більше реклами на відповідну аудиторію, плюс ці дані допоможуть зрозуміти, якою саме це може бути реклама (спрямована більше на чоловіків чи на жінок).
3. Запускати додаткове промо в найменш продуктивні дні, щоб збільшити дохід, підсилити найбільш прибуткові дні (наприклад, додатковий персонал чи акції). На основі цих даних взагалі можна складати графік роботи персоналу - в активні дні більше співробітників на роботі, в менш активні, достатньо 3-4 працівників тощо.